Add support for DeepseekAI's DeepseekVL

geetu040 opened the pull request:
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/36110

This PR adds [DeepseekAI](https://deepseek.ai/)'s DeepseekVL model to Hugging Face Transformers.

DeepseekVL is an open-source Vision-Language (VL) Model designed for real-world vision and language understanding applications. DeepSeek-VL possesses general multimodal understanding capabilities, capable of processing logical diagrams, web pages, formula recognition, scientific literature, natural images, and embodied intelligence in complex scenarios.

Relevant Links
- Research Paper: [DeepSeek-VL: Towards Real-World Vision-Language Understanding](https://arxiv.org/abs/2403.05525)
- Authors: [Haoyu Lu](https://arxiv.org/search/cs?searchtype=author&query=Lu,+H), [Wen Liu](https://arxiv.org/search/cs?searchtype=author&query=Liu,+W), [Bo Zhang](https://arxiv.org/search/cs?searchtype=author&query=Zhang,+B), et al.
- Implementation: [github.com/deepseek-ai/DeepSeek-VL](https://github.com/deepseek-ai/DeepSeek-VL/)
- Models Weights: [huggingface.co/collections/deepseek-ai/deepseek-vl](https://huggingface.co/collections/deepseek-ai/deepseek-vl-65f295948133d9cf92b706d3)

CC: @Benjamin-eecs, @RERV (github contributors of [deepseek-ai/DeepSeek-VL](https://github.com/deepseek-ai/DeepSeek-VL/))

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@ArthurZucker, @Rocketknight1, @Cyrilvallez, @zucchini-nlp 

## TODOs

- [x] Documentation
- [x] Import Statements and Auto Modeling
- [x] Modeling
  - [x] VisionModel
  - [x] TextModel
  - [x] AlignerModel
- [x] Tests
- [x] Integration Tests
- [x] Weight Conversion Script
- [x] Model Cards
- [x] Tokenizer/Processor
- [x] Fix CI/CD tests

geetu040 committed:
upload initial code

Shakib-IO committed:
update deepseek-vl adaptor

geetu040 committed:
update hierarchy of vision model classes

geetu040 committed:
udpate aligner model

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
add text model

Shakib-IO committed:
Added Image Processor

Shakib-IO committed:
Added Image Processor

Shakib-IO committed:
Added Image Processor

geetu040 committed:
apply masks

geetu040 committed:
Merge remote-tracking branch 'fork/deepseek-vl' into deepseek-vl

geetu040 committed:
remove projection; add aligner

geetu040 committed:
remove interpolate_pos_encoding

geetu040 committed:
remove unused params in config

geetu040 committed:
cleaning

geetu040 commented:
@zucchini-nlp , @Rocketknight1, @Cyrilvallez

The `Deepseek-VL` uses `Sam` as backbone for encoding high-resolution images.
And to be more specific, the backbone is `SamVisionEncoder` instead of `SamModel`, which is not available as a public class. By which I mean that, you can do following with `SamModel` but not with `SamVisionEncoder`
```py
from transformers import SamConfig, SamModel
config = SamConfig()
model = SamModel(config)
```
---
I think that we should rename `SamVisionEncoder` -> `SamVisionModel`, inherit it from `SamPreTrainedModel` and make it accessible to the user. I don't think it breaks backward compatibility in any way.

Otherwise, we would have to copy all the classes that build `SamVisionEncoder` for deepseek. There is nothing wrong with this either but having a `SamVisionModel` along with a `SamModel` makes sense, since it might benefit someone else as well.

If you think having a `SamVisionModel` makes sense, should that be done in a separate PR?

Btw, final results would look like this
```py
from transformers import SamVisionConfig, SamVisionModel
config = SamVisionConfig()
model = SamVisionModel(config)
```
and `SamVisionConfig` is already available publically.

zucchini-nlp commented:
@geetu040 we had similar situation with ideficsVision afair. Yes, in that case, we can just make it public and add in the docs. Renaming though would be breaking, imo we can leave name as is

geetu040 commented:
@zucchini-nlp is it okay to do it in the same PR? or should I create a new one

zucchini-nlp commented:
@geetu040 imo a new PR will make it easier for us to iterate and review

geetu040 commented:
Hi @zucchini-nlp, I am working on the `SamVisionEncoder` (going to create the PR soon) and I have a quick question.
I realized that `SamVisionAttention` and `SamVisionSdpaAttention` produce `attn_weights` of different shapes when `output_attentions=True`.

Can you please answer these 2 questions:
1. Is this allowed in `transformers` for the 2 attentions to produce outputs of different shapes?
2. And lets suppose we do something that changes the shape of `output_attentions`, does that break backward compatibility?

zucchini-nlp commented:
@geetu040 no, that is not expected to have different shapes. Usually using sdpa attention means that no `attn_weights` are returned, so it should be available only through 'eager' attention modules

I see that the weights are calculated on top of SDPA by manual matmul of key and query, which imo defeats the purpose of using SDPA in the first place. Can you remove the returned attention and raise warning similar to what is done in ViT?

geetu040 commented:
@zucchini-nlp sure I'll do that.

Shakib-IO committed:
Add the __init__ file

Shakib-IO committed:
added processing deepseek_vl class

Shakib-IO committed:
modified the deepseek-vl processor

Shakib-IO committed:
modified the deepseek-vl processor

Shakib-IO committed:
update __init__

Shakib-IO committed:
Update the image processor class name

Shakib-IO committed:
Added Deepseek to src/transformers/__init__.py file

Shakib-IO committed:
Added Deepseek to image_processing_auto.py

Shakib-IO committed:
update the __init__ file

Shakib-IO committed:
update deepseek_vl image processor

Shakib-IO committed:
Update Deepseek Processor

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
Merge remote-tracking branch 'fork/deepseek-vl' into deepseek-vl

geetu040 committed:
upload fast image processor

geetu040 committed:
Revert "upload fast image processor"

This reverts commit 68c8fd50bafbb9770ac70c9de02448e2519219b4.

geetu040 committed:
update image processor

geetu040 commented:
Hi @zucchini-nlp I have a question regarding the config and model hierarchy, considering how the sub_configs are handled in transformers by `PreTrainedModel` and the tests. Which one would you prefer? I have shown the configuration below, the same would be reflected for the models.

(1)
```
config: DeepseekVLConfig
    text_config: LlamaConfig
    vision_config: DeepseekVLVisionConfig
        use_high_res: bool
        low_res_config: SiglipVisionConfig
        high_res_config: SamVisionConfig
```
(2)
```
config: DeepseekVLConfig
    text_config: LlamaConfig
    use_high_res: bool
    low_res_vision_config: SiglipVisionConfig
    high_res_vision_config: SamVisionConfig
```
Or if there is a third better solution.

---

CC: @Shakib-IO

zucchini-nlp commented:
@geetu040 the second one is more aligned with transformers format. The main thing to follow, is to store all values used in a pre-trained backbone in its own config. Then all these backbone configs need to be under a general `MyModelConfig`. If there are values used outside of the backbone, like `mm_proj_dimension`, we add them under the big general config

geetu040 committed:
flatten heirarchy

geetu040 committed:
remove DeepseekVLModel

geetu040 committed:
major update (complete modeling)

geetu040 committed:
auto modeling and other files

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
formatting

geetu040 committed:
fix quality

geetu040 committed:
replace torchvision in modeling

geetu040 committed:
set default do_normalize to False

geetu040 committed:
add fast image processor template using tool

geetu040 committed:
update image processors

geetu040 committed:
add fast image processor to other files

geetu040 committed:
update liscense

Shakib-IO committed:
Added deepseek image testcases

Shakib-IO committed:
update image test

geetu040 committed:
update processor

geetu040 committed:
write CHAT_TEMPLATE

geetu040 committed:
update model for processor

geetu040 committed:
fix processor

geetu040 committed:
minor fixes and formatting

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
fix image processing and tests

geetu040 committed:
fix interpolation in sam

geetu040 committed:
fix output_attentions in DeepseekVLModel

geetu040 committed:
upload test_modeling

geetu040 committed:
fix tests because of vocab size

geetu040 committed:
set use_high_res_vision=False in tests

geetu040 committed:
fix all modeling tests

geetu040 committed:
fix styling

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
remove explicit background_color from image processors

Shakib-IO committed:
added test_processor

Shakib-IO committed:
added test_processor

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
fix processor tests

geetu040 committed:
update docs

geetu040 committed:
update docs

geetu040 committed:
update docs

geetu040 commented:
@zucchini-nlp , @Rocketknight1, @Cyrilvallez

Hi everyone, I've moved this PR out of draft. Everything is complete except for the model cards. It's now ready for review. Thanks!

geetu040 committed:
update conversion script

Shakib-IO committed:
Fixed typos

zucchini-nlp commented on src/transformers/models/deepseek_vl/modeling_deepseek_vl.py:
```diff
@@ -0,0 +1,573 @@
+            image_features = image_features.to(inputs_embeds.device, inputs_embeds.dtype)
+            inputs_embeds = inputs_embeds.masked_scatter(image_attention_mask, image_features)
+
+        return self.language_model(
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            cache_position=cache_position,
+            flash_attn_kwargs=flash_attn_kwargs,
+        )
+
```
We should return `image_hidden_states` at output, similar to Llava. Same for conditional generation

zucchini-nlp commented on src/transformers/models/deepseek_vl/modeling_deepseek_vl.py:
```diff
@@ -0,0 +1,573 @@
+    "The bare DeepseekVL Model outputting raw hidden-states without any specific head on top.",
+    DEEPSEEK_VL_START_DOCSTRING,
+)
+class DeepseekVLModel(DeepseekVLPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+        self.use_high_res_vision = config.use_high_res_vision
+
+        self.low_res_vision_encoder = DeepseekVLSiglipVisionEncoder(config.low_res_vision_config)
+        if self.use_high_res_vision:
+            output_size = config.low_res_vision_config.image_size // config.low_res_vision_config.patch_size
+            self.high_res_vision_encoder = DeepseekVLSamVisionEncoder(
+                config.high_res_vision_config, output_size=output_size
+            )
+
+        self.language_model = LlamaModel(config.text_config)
+        self.aligner = DeepseekVLAligner(config)
+
+        self.gradient_checkpointing = False
+        # Initialize weights and apply final processing
+        self.post_init()
+
```
looks really cool, but for the backbones we need to use a `PreTrainedModel` and init them with `AutoModel.from_config(text_config/vision_config)`. That is needed for correct attention implementation to be used and allows part of DeepSeek to be init separately

What I suggest is:
- in `DeepSeekVLModel` use AutoModel and init SAM/Llama/Siglip bare backbones
- `get_image_features` method does the extra processing before and after Siglip + applies proj
- `get_hres_image_features` does same for SAM
- remove `DeepseekVLSiglipVisionEncoder` and `DeepseekVLSAMVisionEncoder`

zucchini-nlp commented on src/transformers/models/deepseek_vl/processing_deepseek_vl.py:
```diff
@@ -0,0 +1,160 @@
+        one_img_tokens = self.image_token * self.num_image_tokens
+        for prompt in text:
+            prompt = prompt.replace(self.image_token, one_img_tokens)
+            if self.use_default_system_prompt:
+                prompt = DEFAULT_SYSTEM_PROMPT + prompt
```
system prompt should go in default chat template

zucchini-nlp commented on src/transformers/models/deepseek_vl/processing_deepseek_vl.py:
```diff
@@ -0,0 +1,160 @@
+    tokenizer_class = ("LlamaTokenizer", "LlamaTokenizerFast")
+
+    def __init__(self, image_processor, tokenizer, chat_template=None, use_default_system_prompt=True, **kwargs):
+        self.num_image_tokens = 576
```
can make it configurable through processor's config file

zucchini-nlp commented on src/transformers/models/deepseek_vl/processing_deepseek_vl.py:
```diff
@@ -0,0 +1,160 @@
+
+    attributes = ["image_processor", "tokenizer"]
+    valid_kwargs = ["chat_template", "use_default_system_prompt"]
+    image_processor_class = ("DeepseekVLImageProcessor", "DeepseekVLImageProcessorFast")
+    tokenizer_class = ("LlamaTokenizer", "LlamaTokenizerFast")
```
nit: `AutoImageProcessor` and `AutoTokenizer` will work also

zucchini-nlp commented on tests/models/deepseek_vl/test_image_processing_deepseek_vl.py:
```diff
@@ -0,0 +1,120 @@
+# coding=utf-8
+# Copyright 2021 HuggingFace Inc.
```
2025

zucchini-nlp reviewed (commented):
Very clean work, thanks for adding the model! 💛 

Overall looks very much ready for core maintainer's review. Only thing I want to update is the modeling part with Auto backbones, left a comment below 👇🏻  

Also we could use modular, to copy parts of processing and image processing. Not required though

geetu040 committed:
minor fixes from review

- remove model_id comments in examples
- remove from pre-trained auto mapping
- move to image-text-to-text from vision-to-seq in auto mapping
- add image_token_index to __init__ for config
- remove outdated temporary config in conversion script
- update example to use chat_template in docstring example
- update liscense 2021->2025

geetu040 committed:
fix type in config docstring

Co-authored-by: Raushan Turganbay <raushan.turganbay@alumni.nu.edu.kz>

geetu040 committed:
update get_image_features

geetu040 committed:
fix config

geetu040 committed:
improve DeepseekVLImageProcessor.preprocess

geetu040 committed:
return image_hidden_states

geetu040 committed:
use AutoTokenizer and AutoImageProcessor in Processor

geetu040 committed:
fix model outputs

geetu040 committed:
make num_image_tokens configurable

geetu040 committed:
fix docstring of processor

geetu040 committed:
move system prompt to chat template

geetu040 commented on src/transformers/models/deepseek_vl/modeling_deepseek_vl.py:
```diff
@@ -0,0 +1,573 @@
+        self.global_attn_index = config.global_attn_indexes[0]
+
+        self.model = SamVisionEncoder(config)
+        self.global_neck = deepcopy(self.model.neck)
```
I saw this in the original implementation that they apply the copied neck on the selected hidden state as well, and these weights are not shared
For Reference: [deepseek-ai/DeepSeek-VL](https://github.com/deepseek-ai/DeepSeek-VL/blob/main/deepseek_vl/models/sam.py#L165)

geetu040 committed:
Merge branch 'main' into deepseek-vl

zucchini-nlp commented on src/transformers/models/deepseek_vl/modeling_deepseek_vl.py:
```diff
@@ -0,0 +1,677 @@
+            "low_res_vision_std", torch.tensor(IMAGENET_STANDARD_STD).reshape(1, 3, 1, 1), persistent=False
+        )
+
+        if self.use_high_res_vision:
+            self.output_size = config.low_res_vision_config.image_size // config.low_res_vision_config.patch_size
+            self.global_attn_index = config.high_res_vision_config.global_attn_indexes[0]
```
Another concern about adding a whole backbone depending on a flag, it's not the first time I'm seeing this trend in VLMs

An option more aligned with transformers but with code duplication can be to have two separate `DeepSeekVisionEncoder` for Siglip and `DeepSeekVisionEncoderHighRes` for Siglip+Sam. Then here we add one of two depending on config. Leaving open for discussion as well

geetu040 replied:
What do you think about having 2 completely separate models, removing the need for a flag
- `DeepseekVL`
   - for checkpoint: [deepseek-ai/deepseek-vl-1.3b-chat](https://huggingface.co/deepseek-ai/deepseek-vl-1.3b-chat)
   - uses Llama + SigLip
- `DeepseekVLHybrid`
  - for checkpoint: [deepseek-ai/deepseek-vl-7b-chat](https://huggingface.co/deepseek-ai/deepseek-vl-7b-chat)
  - uses Llama + Siglip + Sam

zucchini-nlp replied:
yeah, that also sounds good to me

ArthurZucker replied:
this sounds better! ✔️

zucchini-nlp reviewed (approved):
Great! I think we can add SamNect with "Copied from" which is more readable than `copy.deepcopy(mylayer)`

Overall LGTM, approving. I left two open to discussion comments, for core maintainer's decision 😉

geetu040 committed:
Merge branch 'main' into deepseek-vl (fix conflicts with DeepseekV3)

geetu040 committed:
fix repo consistency

geetu040 committed:
fix return_dict

geetu040 committed:
replace SamVisionEncoder with SamVisionModel

geetu040 committed:
update to remove deepcopy

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 commented on src/transformers/models/deepseek_vl/modeling_deepseek_vl.py:
```diff
@@ -0,0 +1,677 @@
+        return output
+
+    def get_low_res_image_features(self, pixel_values):
+        pixel_values = F.interpolate(
+            pixel_values.float(),
+            size=self.config.low_res_vision_config.image_size,
+            mode="bilinear",
+            antialias=True,
+        )
+        pixel_values = (pixel_values - self.low_res_vision_mean) / self.low_res_vision_std
```
cc: @ArthurZucker 

The DeepseekVLProcessor resizes the image to 1024 without normalization. The DeepseekVLModel uses SigLip and Sam for encoding the image.
- It keeps the image at 1024 normalized with OpenAI Clip mean/std before giving to Sam
- and resizes the image to 224 normalized with ImageNet Standard mean/std before giving to SigLip

We want to move all this processing to the `DeepseekVLProcessor`, which returns 2 images as `pixel_values` and `pixel_values_high_res`

For reference in code: [DeepseekVLModel](https://github.com/huggingface/transformers/blob/f5e46e4b9cb59638bf5a6e49a1137527dc875463/src/transformers/models/deepseek_vl/modeling_deepseek_vl.py#L378)

ArthurZucker replied:
Yes! This sounds better! Let's avoid any image processing in the modeling 🤗

geetu040 committed:
Merge branch 'main' into deepseek-vl

ArthurZucker commented on src/transformers/models/deepseek_vl/modeling_deepseek_vl.py:
```diff
@@ -0,0 +1,727 @@
+        # interpolate Sam encodings to match Siglip encodings
+        features = F.interpolate(
+            features.float(),
+            size=(4 * self.output_size, 4 * self.output_size),
```
this could / should be a config param no?

ArthurZucker reviewed (commented):
Answered questions, LGTM in general good work and good direction!

geetu040 committed:
Merge branch main into deepseek-vl (Resolve Conflicts)

geetu040 committed:
🛠️  Major Architectural Changes (Adds DeepseekVLHybrid)

geetu040 committed:
fix quality checks

geetu040 committed:
add missing hybrid in auto modeling

geetu040 committed:
Merge branch 'main' into deepseek-vl

ArthurZucker commented:
Feel free to ping me again once ready!

geetu040 committed:
Merge branch main into deepseek-vl (resolve conflicts)

geetu040 committed:
run make style

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
update sam_hq

geetu040 committed:
update high_res_size in test

geetu040 committed:
update docs following #36979

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
update code with auto_docstring

geetu040 committed:
update conversion scripts

geetu040 committed:
fix style

geetu040 committed:
fix failing test because of tuple

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
set weights_only=True in conversion script

geetu040 committed:
use safetensors.torch.load_file instead of torch.load in conversion script

geetu040 committed:
make output_dir optional in conversion script

geetu040 committed:
fix code snippets in docs (now the examples work fine)

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
integration tests for DeepseekVL

geetu040 committed:
update expected texts

geetu040 committed:
make style

geetu040 committed:
integration tests for DeepseekVLHybrid

geetu040 committed:
fix class name

geetu040 committed:
update expected texts for hybrid

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 commented:
This PR is now 100% complete, covering modeling, processing, documentation, checkpoint conversion, tests, and model cards. It should be ready for review.

I have also transferred the checkpoints from my personal Hugging Face account [geetu040](https://huggingface.co/geetu040) to the [deepseek-community](https://huggingface.co/deepseek-community) account. You can find these updated checkpoints in this collection: [Deepseek-VL](https://huggingface.co/collections/deepseek-community/deepseek-vl-684d369d6e04c26577750485)
- [geetu040/deepseek-vl-1.3b-chat](https://huggingface.co/geetu040/deepseek-vl-1.3b-chat)
- [geetu040/deepseek-vl-1.3b-base](https://huggingface.co/geetu040/deepseek-vl-1.3b-base)
- [geetu040/deepseek-vl-7b-chat](https://huggingface.co/geetu040/deepseek-vl-7b-chat)
- [geetu040/deepseek-vl-7b-base](https://huggingface.co/geetu040/deepseek-vl-7b-base)

---

@zucchini-nlp previously reviewed and approved the modeling code in [https://github.com/huggingface/transformers/pull/36248#pullrequestreview-2725116254](https://github.com/huggingface/transformers/pull/36248#pullrequestreview-2725116254). Since then, I haven’t changed the core implementation, just mainly refactored by separating `DeepseekVLHybrid` from `DeepseekVL`, as suggested here: [https://github.com/huggingface/transformers/pull/36248#discussion\_r2018434624](https://github.com/huggingface/transformers/pull/36248#discussion_r2018434624).

CC: @ArthurZucker, @LysandreJik, @zucchini-nlp, @Rocketknight1

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
run "make style"

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
update since changes in main

ArthurZucker commented:
Sorry sir! Reviewing

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
run make-style

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
Merge branch main into deepseek-vl (resolve conflicts)

geetu040 committed:
nits since changes in main

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
undo changes in sam

geetu040 committed:
fix tests

geetu040 committed:
fix tests; update with main

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
update with main: output_attention/output_hidden_states

geetu040 committed:
fix copied part in deepseek_vl

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
run fix-copies

geetu040 committed:
fix output_hidden_states

geetu040 committed:
sam: fix _init_weigths

geetu040 commented:
#### Notes regarding SAM
I have changed the `_init_weights` for `SamPreTrainedModel` which also reflects changes in `SamHQPreTrainedModel`. This was intentional and required because of the failing tests at [jobs/1817956](https://app.circleci.com/pipelines/github/huggingface/transformers/137157/workflows/5d6721a9-c434-48dc-83cb-d32ea4a1fe4e/jobs/1817956)
```
FAILED tests/models/deepseek_vl_hybrid/test_modeling_deepseek_vl_hybrid.py::DeepseekVLHybridModelTest::test_can_init_all_missing_weights - AssertionError: False is not true : The following keys are not properly handled by `_init_weights()`:
['high_res_vision_model.vision_encoder.pos_embed']
```
Also, this makes sense, since `pos_embed` had already been explicitly set to zero for `SamHQPreTrainedModel`, so doing same for `SamPreTrainedModel` doesnot hurt anything.
If you think this should have been done in a separate PR, let me know, I'll create one!

geetu040 committed:
Merge branch 'main' into deepseek-vl

ArthurZucker commented on src/transformers/models/deepseek_vl/configuration_deepseek_vl.py:
nice and simple 🧼

ArthurZucker commented on src/transformers/models/deepseek_vl/modeling_deepseek_vl.py:
```diff
@@ -0,0 +1,361 @@
+            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.text_config.vocab_size)
+
+        return DeepseekVLCausalLMOutputWithPast(
+            loss=loss,
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            image_hidden_states=outputs.image_hidden_states,
+        )
+
+    # Copied from transformers.models.janus.modeling_janus.JanusForConditionalGeneration.prepare_inputs_for_generation with Janus->DeepseekVL
+    def prepare_inputs_for_generation(
+        self,
+        input_ids,
+        pixel_values=None,
+        past_key_values=None,
+        attention_mask=None,
+        inputs_embeds=None,
+        cache_position=None,
+        logits_to_keep=None,
+        **kwargs,
+    ):
+        # Overwritten -- extra custom processing
+
+        model_inputs = super().prepare_inputs_for_generation(
+            input_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            attention_mask=attention_mask,
+            cache_position=cache_position,
+            logits_to_keep=logits_to_keep,
+            **kwargs,
+        )
+
+        # If we're in cached decoding stage, pixel values should be None because input ids do not contain special image token anymore
+        # Otherwise we need pixel values to be passed to model
+        if cache_position[0] == 0:
+            model_inputs["pixel_values"] = pixel_values
+
```
If there is `copied from` it means modular can be used! Let's use modular! 🤗

geetu040 replied:
Fixed in [0a2ff70](https://github.com/huggingface/transformers/pull/36248/commits/0a2ff70b3aaac33165c40e9eae8d6fbac7844e30)

ArthurZucker commented on src/transformers/models/deepseek_vl/image_processing_deepseek_vl.py:
I think a lot of this is already in the lib, if you could just check other `image_processing` to inherit when you use modular!

geetu040 replied:
Fixed in [b88e03f](https://github.com/huggingface/transformers/pull/36248/commits/b88e03fc72496d721555b1ab5f397e5e5ea9f4c0)

ArthurZucker commented on src/transformers/models/deepseek_vl_hybrid/modular_deepseek_vl_hybrid.py:
nice!

ArthurZucker reviewed (commented):
Very big PR! Kudos to you 🤗

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
use modular for DeepseekVL

geetu040 committed:
make image processor more modular

zucchini-nlp commented on src/transformers/models/deepseek_vl/modeling_deepseek_vl.py:
```diff
@@ -0,0 +1,351 @@
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["LlamaDecoderLayer"]
+    _skip_keys_device_placement = ["past_key_values", "causal_mask"]
+    _supports_flash_attn_2 = True
+    _supports_sdpa = True
+    _supports_quantized_cache = True
+    _supports_cache_class = True
```
can remove `cache` flags and rename `flash_attn_2` to `flash_attn` 😄 There was a major clean up recently

geetu040 replied:
Fixed in [91295c2](https://github.com/huggingface/transformers/pull/36248/commits/91295c24f7379f7c5c611f6ca1021a7ed7afc2d9)

geetu040 committed:
Merge branch 'main' into deepseek-vl

zucchini-nlp commented on src/transformers/models/deepseek_vl/modeling_deepseek_vl.py:
```diff
@@ -0,0 +1,351 @@
+
+        loss = None
+        if labels is not None:
+            loss = self.loss_function(logits=logits, labels=labels, vocab_size=self.config.text_config.vocab_size)
```
yep, most prob we missed it when reviewing Janus

geetu040 replied:
Fixed in [96862d8](https://github.com/huggingface/transformers/pull/36248/commits/96862d806be83c865f92b9bf1d025974a463101c)

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
modular: use JanusPreTrainedModel

geetu040 committed:
janus: provide kwargs in loss

geetu040 commented on src/transformers/models/deepseek_vl_hybrid/modular_deepseek_vl_hybrid.py:
```diff
@@ -0,0 +1,768 @@
+            do_convert_rgb (`bool`, *optional*, defaults to `self.do_convert_rgb`):
+                Whether to convert the image to RGB.
+        """
+        do_resize = do_resize if do_resize is not None else self.do_resize
+        do_rescale = do_rescale if do_rescale is not None else self.do_rescale
+        do_normalize = do_normalize if do_normalize is not None else self.do_normalize
+        resample = resample if resample is not None else self.resample
+        high_res_resample = high_res_resample if high_res_resample is not None else self.high_res_resample
+        rescale_factor = rescale_factor if rescale_factor is not None else self.rescale_factor
+        image_mean = image_mean if image_mean is not None else self.image_mean
+        image_std = image_std if image_std is not None else self.image_std
+        high_res_image_mean = high_res_image_mean if high_res_image_mean is not None else self.high_res_image_mean
+        high_res_image_std = high_res_image_std if high_res_image_std is not None else self.high_res_image_std
+        do_convert_rgb = do_convert_rgb if do_convert_rgb is not None else self.do_convert_rgb
```
Fixed in [b88e03f](https://github.com/huggingface/transformers/pull/36248/commits/b88e03fc72496d721555b1ab5f397e5e5ea9f4c0)
Used modular for image processor, which removed all these lines.

geetu040 committed:
update processors in conversion script

geetu040 committed:
Merge branch 'main' into deepseek-vl

ArthurZucker commented:
On it!

geetu040 committed:
Revert "sam: fix _init_weigths"

This reverts commit db625d0c68956c0dad45edd7a469b6a074905c27.

geetu040 committed:
Merge branch 'main' into deepseek-vl

geetu040 committed:
run fix-copies

ArthurZucker reviewed (approved):
Let's go thanks for the hardwork and good PR! 🚀
